stdlib/dynarray.mli

```diff
@@ -0,0 +1,190 @@
+(*                            Simon Cruanes                               *)
+(*                                                                        *)
+(*   Copyright 2022 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Growable, mutable array.
+
+
+
+    @since 5.1
+*)
+
+type 'a t
+(** A dynamic array containing values of type ['a].
+
+    This contains an underlying {!array} along with a size.
+    Operations such as {!push}, {!append}, and {!append_seq}, extend the
+    size (and might reallocate the underlying array).
+
+    Operations such as {!pop}, and {!truncate}, reduce the size. *)
+
+val create : unit -> 'a t
+(** Create a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] makes a array of size [n], filled with [x]. *)
+
+val init : int -> (int -> 'a) -> 'a t
+(** Init the array with the given function and size. *)
+
+val clear : 'a t -> unit
+(** Clear the content of the array.
+    This ensures that [length v = 0] but the underlying array is kept,
+    and possibly references to former elements, which are therefore
+    not garbage collectible. *)
+
+val ensure_capacity_with : 'a t -> filler:'a -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size).
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @param filler an element used if the underlying array is empty,
+      to initialize it. It will be retained until the array is totally
+      empty or until it is garbage collected.
+    @raise Invalid_arg if the size is not suitable (negative, or too big for
+    OCaml arrays)
+*)
+
+val ensure_capacity_nonempty : 'a t -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size),
+    assuming it is non-empty. The first element is used as the filler.
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @raise Invalid_arg if the array is empty or
+      if the size is not suitable (negative, or too big for OCaml arrays)
+*)
+
+val is_empty : 'a t -> bool
+(** Is the array empty? This is synonymous to [length a = 0]. *)
+
+val push : 'a t -> 'a -> unit
+(** Add an element at the end of the array. This might extend the underlying
+    array if it is full.
+
+    Calling [push] [n] times is amortized O(n) complexity,
+    and O(ln(n)) reallocations of the underlying array. *)
```

Comment:
I think using `push/pop` for adding/poping at the end is confusing (cue `Queue`). It also doesn't work well or deque like datastructure (e.g. the persistent vectors I was talking about). I would suggest `{add,pop}_last` so that when we get to deque we can use `{add,pop}_{first,last}` and have all that menagerie of data structures relatively consistent.

Also `add` is in fact the word you use in this doc string.

Reply:
C++ uses `push_back`, we could do the same. It's consistent with `Stack.push` and has the same semantics.

Reply:
> C++ uses push_back, we could do the same. It's consistent with Stack.push and has the same semantics.

I think we got mixed up. I meant `Stack` you meant `Queue`. In any case none of these datastructures provide random access to elements and have a clear access policy so bare `push`, `pop` names are clear in their intent.

Here we have an array with random access. When I first saw your `push`, I thought it was adding elements at the front. Regarding `push_back`, it's in my opinion not so great, it again implies a notion of order of access more related to queues.

Since in the end this is an array I think the `first` and `last` elements of an array are pretty clear concepts so I would rather go with `add_last`.

Reply:
I mean, the word "push" is consistent with `Stack`, and a dynarray is a stack. A C++ vector uses `push_back` for that; a rust Vec uses `push`. It's well established. Otoh, java's ArrayList uses `add` which I find very vague. `push_back` has the advantage that it both matches Stack's "push" word, and specifies futher where the item is added because of "_back" (which is _also_ consistent with a Deque since a Deque is also a stack).

For me `add_last` is weird, but I'm open to the suggestion if there is consensus on it.

Reply:
Well now that it seems we have `pop_last`, `pop_last_opt` and `remove_last`, I don't see a good reason why one would want to add a new word to talk about this last element…

Reply:
> and a dynarray is a stack.

Sorry to insist on that one but to me the datastructure you propose in this PR is simply an extensible array. An array to which I can add elements at the end efficiently. Seeing it as a stack is an interpretation. 

When I add an element at the end of this array nothing actually gets pushed, you literally write yourself how it should be read, there's no need to change the verb:

```
[push_last a x] adds the element [x] at the end of the array [a].
```

Not to mention that since we have `{Hashtbl,Map,Set}.{add,remove}`, I don't see a good reason why we should have `{Dynarray.push,remove}_last`.

Reply:
I'm not sure what to say. Dynamic arrays are commonly used as stacks (and I see little reason to use `Stack` personally). Lots of languages have no dedicated stack data structure besides their versions of dynarray. I don't know what "an interpretation" is, when all the contracts required by the notion of a stack are satisfied by a given data structure.

I think `add` is not specific enough because `push_last` adds at the _only_ position where addition is reasonably fast: the end. `add` could as well mean adding at the beginning, or inserting in the middle. `Hashtbl/Set/Map` do not tell you where they put the element, but `push_last` does, and you can then use `pop_last` or the likes.

In fact, we probably need `last_opt : 'a t -> 'a option` and `last : 'a t -> 'a` to go with the current API, like [Vec::last](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last).

Reply:
> I think add is not specific enough because push_last adds at the only position where addition is reasonably fast: the end.

I never suggested `add`, I suggested `add_last` which closely follows what your doc string says.

> In fact, we probably need last_opt : 'a t -> 'a option and last : 'a t -> 'a to go with the current API,

Following `Either` I would suggest `find_last : 'a t -> 'a option` and `get_last : 'a t -> 'a` 
as an alias for `get a (Dynarray.length a - 1)`.

Reply:
I suppose I can live with `add_last`, even if I personally find it terrible. I think `push` is more familiar to a lot of programmers, but well. I'm open to it.


`get_last` sounds fine, I'll add it. I think `find_last` is a mistake: `find`, in every other linear stdlib container, means: "traverse the structure to find an element that satisfy a predicate". `find_last` here would be traversing from the end.

Reply:
I do think that very often dynamic arrays are used as infinite arrays that are appended to at the end (similar to a more efficient backwards list). After filling the array, you may then access or change various parts of it. In this very popular use-case, the usage pattern is similar to that of a stack while still allowing for random access.

One way to generalize this operation is to create an `insert` function, which effectively clears the targeted slot by moving all subsequent entries in the array. `insert array x 0` would therefore 'push' to the front of the array, whereas `insert array x {length of array}` (or simply `insert array x (-1)`) would add to the end of it. An `insert_at_end` function could be a shortcut for this functionality.

Reply:
@dbuenzli I am not familiar with RRB vectors but your point with going straight to a better structure makes a lot of sense. I mentioned the hoops Java went through to make their collections somehow fit for a concurrent setting, and if people do not like what I showed and prefer fewer data-race protections (because the OCaml programmer of course would not write buggy code), then they should think twice about adding dynamic arrays to the stdlib instead of a better structure. It depends on how much "better" the other one is, of course.

Reply:
> It depends on how much "better" the other one is, of course.

For me that's an open question. The RRB vectors look good to me [on paper](https://doi.org/10.1145/2784731.2784739), but there's always a difference between a paper and reality (and different GCs). 

I wouldn't dare saying let's add this to the stdlib without a rigorously benchmarked implementation and a little tour on how it turned out on the scala side 7 years later (I once started an [an incomplete implementation](https://github.com/dbuenzli/pvec/blob/master/src/pvec.mli) because I wanted to use this for a `utext` data structure, I think there are other ones floating in the OCaml space).

Reply:
> I remain extremely doubtful about adding this specialized, append only, concurrency unsafe datastructure after we did without it during last 25 years and with multicore at the corner, but it seems I'm the only one.

I remain confident about adding this general purpose, constant time access, stack-like, extremely fast datastructure. It's not intended for multithreading, so what? Most of the stdlib isn't either. There is no faster structure for accessing items by index, besides `array` (which has the problem that it's more difficult to build one). In a way, dynarray is _more_ general than `array`, and many languages skip `array` altogether.

I agree that `insert` is not a good idea, it hides the cost of inserting in the middle or beginning. It makes it too easy to use a O(n) operation instead of O(1) amortized. I guess I'll use `add_last` and move on to more productive discussions.

As for `find`, I think its use in `Either` is a mistake :slightly_smiling_face: . For me `find_last` for a dynarray would really mean:

```ocaml
val find_last : ('a -> bool) -> 'a t -> int * 'a
(** [find_last p a] finds the last item of [a] that satisfies [p]
    @raise Not_found if no item satisfies [p] *)

(* or *)

val find_last : ('a -> bool) -> 'a t -> 'a
```

Reply:
>  I guess I'll use add_last and move on to more productive discussions.

I'm deeply sorry to hinder your productivity, master. 

> For me find_last for a dynarray would really mean

Because after all these years you still seem to design APIs without reading them in context (you are not alone). That could be a hint to have more productive discussions on your designs and maybe, even better ones…

Reply:
I think there's plenty of context in the stdlib around `find` being associated with a predicate, especially with containers! `Set`, `Map`, `List`, at least. Why should I have less context than you when we both use the stdlib daily?

I also think I've been patient in this PR, and tried to listen to your recommendations. I think it's been reasonably productive so far.

Reply:
Would it make sense to accept that people use different metaphors for the data-structure, and export *both* `add_last` and `push` as aliases of each other? Is it something that both @c-cube and @dbuenzli would be happy with, or on the contrary an idea that you both dislike?

Reply:
> Why should I have less context than you when we both use the stdlib daily?

That's not what I meant. I meant that one should not think about identifiers as floating in space unprefixed, but write code with them and see how it reads. If you write: 

```ocaml 
match Dynarray.find_last a with 
| None -> …
| Some last -> …

match Either.find_left e with 
| None -> …
| Some left -> …
```

It's pretty clear what the function does. I think `find_$(item)` is a fine scheme for looking up something that exists or not, regardless of a predicate. And it's complement `get_$(item)` for looking up things that you know must exist and is a programmer error if it did not (i.e. raises `Invalid_argument`, like `Result.get_ok`, `Array.get`, `Option.get`, etc.).

> Is it something that both @c-cube and @dbuenzli would be happy with, or on the contrary an idea that you both dislike?

It depends whether you want things that are designed or not. Design is about making choices. I don't think users of the `Stdlib` should be subjected to a diversity of names because @c-cube and @dbuenzli cannot agree on one. I think that if you want to use `push` then you should not use `append` because you then get an API with frankenstein terminology (see my message above). 

(And citing programmer familiarity or uses in other languages is not enough to me, there exists very bad names like `filter` which while extremely widespread are in fact very bad names, since they need to be learned; that wouldn't be the case with e.g. `drop` or `keep`)

Reply:
Also I forgot to add that on the top of my mind I must have at least 4-5 internal uses of `Dynarray` in my code and none of these are stacks. 

Mostly imperative heap implementations (add new element at the end + heap up), or gradually building arrays. 

I think that if you want to see them as stacks you can just do a `let push = Dynarray.add_last` in your context.

Reply:
I have plenty of uses for them as stacks (notably in all things SAT-related) but I will live with `add_last`, it's fine.

For `find` I still find it very strange and unsettling.

Reply:
(Another option would be to have a `Dynarray.Stack` submodule that exposes `type 'a t = 'a Dynarray.t` and offers a more stack-like vocabulary.)